### PR TITLE
Added struct and slice for dynamic scaling support

### PR DESCRIPTION
**Pull request by tpuhan**
Things in flush are defaulted to index 0 (meaning the first and only managed stream) and I still haven't built a stream yet. This is just the infrastructure for holding each stream.

### CODE REVIEW

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: Perhaps use a more descriptive name for "ind"? Suggestion: streamIndex

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go:570]: Do we need a mutex to control access to the stream slice? When will the slice potentially get modified?

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: How will the stream index be computed in future? Can we put that logic into a function that simply returns 0 for now?

**Review by agrawal-siddharth — CHANGES_REQUESTED** [review]

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Done.

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go:570]: The slice gets modified when we add a new stream to the slice. When I implement this change, I will be adding a mutex around the insertion of the new stream into the slice!

Comment (tpuhan) [reply out_writeapi.go:570]: The slice also gets modified in init, but we only append one stream to it, so putting a mutex there will be unnecessary.

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Done.

File: out_writeapi.go
Comment (agrawal-siddharth) [reply out_writeapi.go:570]: Note that we will need to protect access to the slice with a mutex anywhere it is accessed, whether for read or for write.

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: Perhaps also include stream index in the error message?

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: This code is not rebased to main.

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: Might we select a different stream across chunks? Alternatively we could select the stream up-front and use it for both chunking and non-chunking cases. What are the tradeoffs?

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: No longer need this comment

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: No longer need this comment

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: No longer need this comment

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: No longer need this comment

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: Why do we need this? Don't we have to iterate over all streams in the slice?

**Review by agrawal-siddharth — CHANGES_REQUESTED** [review]

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Done.

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Not yet. I was going to do it after the request got approved, but I can do it now!

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: I believe we can select different streams across chunks in this current implementation. I believe this would be especially useful in cases where there is a lot of input data that is being chunked into a lot of smaller pieces. If we picked a stream up front and used it for the chunking and non-chunking portions, there is a chance that one stream might still have the same issue of back pressure due to AppendRows being called in a for loop. By selecting a different stream across chunks, we can take advantage and lower the amount of requests in a single stream.

Comment (tpuhan) [reply out_writeapi.go]: A trade off would be this makes the ability to guarantee Exactly Once a bit trickier, but not impossible.

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Done.

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Done.

File: out_writeapi.go
Comment (agrawal-siddharth) [reply out_writeapi.go]: In practice, our default chunk size is 9MB and Fluentbit sends us data of at most 2MB in a single call to Flush. Hence chunking will typically never happen. The code would be somewhat simpler if we identified the target stream up-front, because then we need to call this logic in only 1 place rather than in 2 places.

Exactly-once is managed per stream so will work with both approaches.

Again, either approach will work.

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: I'm wondering if we should hold off on returning the error until we are done closing all the streams. So, just set a flag that we encountered this error, but then return the error after the loop is done.

**Review by agrawal-siddharth — CHANGES_REQUESTED** [review]

File: out_writeapi.go
Comment (tpuhan) [reply out_writeapi.go]: Sounds good and done!

File: out_writeapi.go
Comment (agrawal-siddharth) [inline out_writeapi.go]: This check should go after we close the client as well, below.

**Review by agrawal-siddharth — CHANGES_REQUESTED** [review]

**Review by agrawal-siddharth — APPROVED** [review]

**Pull request merged**
